ml-agents/mlagents/trainers/trainer/rl_trainer.py

```diff
@@ -150,6 +150,8 @@ def _maybe_save_model(self, step_after_process: int) -> None:
         if step_after_process >= self._next_save_step and self.get_step != 0:
             logger.info(f"Checkpointing model for {self.brain_name}.")
             self.save_model(self.brain_name)
+            logger.info(f"Exporting a checkpoint model for {self.brain_name}.")
+            self.export_model(self.brain_name, is_checkpoint=True)
```

Comment:
So we have two methods now, `save_model` and `export_model`.

In `save_model` we tell the policy to save its model (saves a checkpoint at a certain # steps).
In `export_model` we save a frozen model and convert it to .nn.

In my opinion it seems clearer to have `checkpoint_model` for checkpoints and `save_model` for the final model.  Both would use our model serialization logic to write the file out to disk.

Curious for your opinion @ervteng -- also curious about whether that logic should actually live in Trainer or Policy.

Reply:
I am not sure I understand the train of thought here. Is it to write out another function (such as `checkpoint_model` as mentioned) which would call `save_model` + `export_model` instead of making the two separate calls in the code?

Reply:
I think the issue is that the API here has become confusing.  Previously `save_model` saved the state of the model as a checkpoint and Tensorflow protobuf, while `export_model` exported the model as a `.nn`.  Now the naming is confusing, IMO, and we actually even have the `is_checkpoint` flag.

The suggestion is to solve this by having one `checkpoint_model` method for saving checkpoints and one `save_model` method for saving the final model -- and I think you understood correctly I meant the checkpoint model would write both the TF checkpoint and the .nn checkpoint.

Reply:
I also don't like how we have to propagate the `is_checkpoint` all the way down. I actually think this needs to be refactored significantly, possibly not in this PR. IMO the Policy should be responsible for exporting itself, as the Trainer basically shouldn't be aware of the implementation (tensorflow or PyTorch, for instance) under the hood. So it shouldn't grab `policy.sess` or `policy.graph`, for instance. 

What about this:
- `Policy` has an abstract method `export_model(path)`
- `TFPolicy` has the concrete implementation of this, which calls `export_policy_model`. 
- `_checkpoint_model` is an internal trainer method that saves the checkpoint and calls `export_model` on its policy
- `save_model` is a public trainer method called by TrainerController and calls `export model` but with a different path. 

Alternatively, we remove the `save_model` method altogether and instead call `trainer.get_policy(brain_name).export_policy` in TrainerController.

Reply:
Saving checkpoints is handled in rl_trainer, but the final model (or on interrupt) is handled in trainer_controller. Within rl_trainer, this is a trivial fix, but the final_model would have to be addressed a little differently.

Ideally, if `checkpoint_saving` and `final_model_save` were happening at the same level, it would be straightforward. I can do this, but it seems like it would expand this PR and increase its scope, as Ervin mentionned.